Subject:
[Qemu-devel] sh4: r2d --append option support
From:
takasi-y@ops.dti.ne.jp
Date:
Wed, 11 Feb 2009 22:15:52 +0900 (JST)
To:
qemu-devel@nongnu.org
X-Account-Key:
account2
X-UIDL:
ee9fb39dacbfe0a513c706a0cdf320bb
X-Mozilla-Status:
0007
X-Mozilla-Status2:
00000000
X-DTI-Recipient:
<kawasaki@juno.dti.ne.jp>
Return-Path:
<qemu-devel-bounces+kawasaki=juno.dti.ne.jp@nongnu.org>
Received:
from lists.gnu.org (lists.gnu.org [199.232.76.165]) by pop.juno.dti.ne.jp (3.10p) with ESMTP id n1BDgN5R016246 for <kawasaki@juno.dti.ne.jp>; Wed, 11 Feb 2009 22:42:24 +0900 (JST)
Received:
from localhost ([127.0.0.1]:35646 helo=lists.gnu.org) by lists.gnu.org with esmtp (Exim 4.43) id 1LXFCO-0004rJ-1W for kawasaki@juno.dti.ne.jp; Wed, 11 Feb 2009 08:32:20 -0500
Received:
from mailman by lists.gnu.org with tmda-scanned (Exim 4.43) id 1LXEwa-0000iY-Q3 for qemu-devel@nongnu.org; Wed, 11 Feb 2009 08:16:00 -0500
Received:
from exim by lists.gnu.org with spam-scanned (Exim 4.43) id 1LXEwZ-0000hu-5R for qemu-devel@nongnu.org; Wed, 11 Feb 2009 08:16:00 -0500
Received:
from [199.232.76.173] (port=42037 helo=monty-python.gnu.org) by lists.gnu.org with esmtp (Exim 4.43) id 1LXEwY-0000hp-VP for qemu-devel@nongnu.org; Wed, 11 Feb 2009 08:15:59 -0500
Received:
from smtp09.dti.ne.jp ([202.216.231.184]:45954) by monty-python.gnu.org with esmtp (Exim 4.60) (envelope-from <takasi-y@ops.dti.ne.jp>) id 1LXEwY-0002tv-B7 for qemu-devel@nongnu.org; Wed, 11 Feb 2009 08:15:58 -0500
Received:
from ubuny (KHP059140016099.ppp-bb.dion.ne.jp [59.140.16.99]) by smtp09.dti.ne.jp (3.11s) with ESMTP AUTH id n1BDFqfq024928; Wed, 11 Feb 2009 22:15:53 +0900 (JST)
Message-ID:
<200902111315.n1BDFqfq024928@smtp09.dti.ne.jp>
X-Mailer:
Sylpheed 2.5.0 (GTK+ 2.14.4; i486-pc-linux-gnu)
MIME-Version:
1.0
Content-Type:
text/plain; charset=US-ASCII
Content-Transfer-Encoding:
7bit
X-detected-operating-system:
by monty-python.gnu.org: Solaris 9
X-BeenThere:
qemu-devel@nongnu.org
X-Mailman-Version:
2.1.5
Precedence:
list
Reply-To:
qemu-devel@nongnu.org
List-Id:
qemu-devel.nongnu.org
List-Unsubscribe:
<http://lists.nongnu.org/mailman/listinfo/qemu-devel>, <mailto:qemu-devel-request@nongnu.org?subject=unsubscribe>
List-Archive:
<http://lists.gnu.org/pipermail/qemu-devel>
List-Post:
<mailto:qemu-devel@nongnu.org>
List-Help:
<mailto:qemu-devel-request@nongnu.org?subject=help>
List-Subscribe:
<http://lists.nongnu.org/mailman/listinfo/qemu-devel>, <mailto:qemu-devel-request@nongnu.org?subject=subscribe>
Sender:
qemu-devel-bounces+kawasaki=juno.dti.ne.jp@nongnu.org
Errors-To:
qemu-devel-bounces+kawasaki=juno.dti.ne.jp@nongnu.org

Add linux kernel command line ("--append" option) support.
Fix kernel loading address to appropriate position when --append used.
Using --kernel but --append case is left untouched for backward compatibility.

Signed-off-by: Takashi YOSHII <takasi-y@ops.dti.ne.jp>
---
I found this has not been posted as a patch to apply.
I hope this be merged into the branch(if any:) for next stable release.
/yoshii
---
 hw/r2d.c |   13 +++++++++----
 1 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/hw/r2d.c b/hw/r2d.c
index e6b2bf4..7dcb723 100644
--- a/hw/r2d.c
+++ b/hw/r2d.c
@@ -233,20 +233,25 @@ static void r2d_init(ram_addr_t ram_size, int vga_ram_size,
         pci_nic_init(pci, &nd_table[i], -1, "ne2k_pci");
 
     /* Todo: register on board registers */
-    {
+    if (kernel_filename) {
       int kernel_size;
       /* initialization which should be done by firmware */
       stl_phys(SH7750_BCR1, 1<<3); /* cs3 SDRAM */
       stw_phys(SH7750_BCR2, 3<<(3*2)); /* cs3 32bit */
 
-      kernel_size = load_image(kernel_filename, phys_ram_base);
+      if (kernel_cmdline) {
+          kernel_size = load_image(kernel_filename, phys_ram_base + 0x80000);
+          env->pc = (SDRAM_BASE + 0x80000) | 0xa0000000;
+          pstrcpy(phys_ram_base + 0x10100, 256, kernel_cmdline);
+      } else {
+          kernel_size = load_image(kernel_filename, phys_ram_base);
+          env->pc = SDRAM_BASE | 0xa0000000; /* Start from P2 area */
+      }
 
       if (kernel_size < 0) {
         fprintf(stderr, "qemu: could not load kernel '%s'\n", kernel_filename);
         exit(1);
       }
-
-      env->pc = SDRAM_BASE | 0xa0000000; /* Start from P2 area */
     }
 }
 
-- 1.5.6.3 